c8d/prune: Keep the last tagged image instead of creating dangling image

vvoland opened the pull request:
- fixes: https://github.com/moby/moby/issues/48063

Don't turn images into dangling when they are used by containers created with an image specified by an ID only (e.g. `docker run 82d1e9d`).

Keep the last image reference with the same target when all other references would be pruned.

If the container was created with a digested and tagged reference (e.g. `docker run alpine:latest@sha256:82d1e9d7ed48a7523bdebc18cf6290bdb97b82302a8a9c27d4fe885949ea94d1`), the `alpine:latest` image won't get untagged.

This change makes the behavior consistent with the graphdriver implementation.

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**
TestPruneDontDeleteUsedImage

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section:
-->
```markdown changelog
containerd integration: Fix `docker image prune -a` untagging images used by containers started from images referenced by a digested reference.
```

**- A picture of a cute animal (not mandatory but encouraged)**

vvoland committed:
c8d/prune: Keep the last tagged image instead of creating dangling image

Don't turn images into dangling when they are used by containers created
with an image specified by an ID only (e.g. `docker run 82d1e9d`).

Keep the last image reference with the same target when all other
references would be pruned.

If the container was created with a digested and tagged reference (e.g.
`docker run alpine:latest@sha256:82d1e9d7ed48a7523bdebc18cf6290bdb97b82302a8a9c27d4fe885949ea94d1`),
the `alpine:latest` image won't get untagged.

This change makes the behavior consistent with the graphdriver
implementation.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d/prune: Add OTEL spans

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented:
Hmm, getting weird failures to start secondary daemon (overlay2 not supported?) in integration tests:
```

2024-06-27T16:05:08.9359404Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.344847699Z" level=info msg="Starting up"
2024-06-27T16:05:08.9362497Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.345499908Z" level=debug msg="Listener created for HTTP on unix (/tmp/docker-integration/df78e0f4df204.sock)"
2024-06-27T16:05:08.9364782Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.354504908Z" level=debug msg="Golang's threads limit set to 115020"
2024-06-27T16:05:08.9366455Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.354997418Z" level=debug msg="metrics API listening on /tmp/dxr/df78e0f4df204/metrics.sock"
2024-06-27T16:05:08.9367646Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.357471248Z" level=debug msg="Using default logging driver json-file"
2024-06-27T16:05:08.9368878Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.357530727Z" level=debug msg="processing event stream" module=libcontainerd namespace=df78e0f4df204p
2024-06-27T16:05:08.9371322Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.357662128Z" level=debug msg="No quota support for local volumes in /go/src/github.com/docker/docker/bundles/test-integration/TestPruneDontDeleteUsedImage/single_tag/repo_and_digest,_no_tag/df78e0f4df204/root/volumes: Filesystem does not support, or has not enabled quotas"
2024-06-27T16:05:08.9373348Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.363082482Z" level=info msg="[graphdriver] trying configured driver: overlay2"
2024-06-27T16:05:08.9374572Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.380383275Z" level=error msg="failed to mount overlay: invalid argument" storage-driver=overlay2
2024-06-27T16:05:08.9375843Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.380407951Z" level=debug msg="daemon configured with a 15 seconds minimum shutdown timeout"
2024-06-27T16:05:08.9377146Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.380416497Z" level=debug msg="start clean shutdown of all containers with a 15 seconds timeout..."
2024-06-27T16:05:08.9378309Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.381139810Z" level=debug msg="Cleaning up old mountid : start."
2024-06-27T16:05:08.9379322Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.381418636Z" level=debug msg="Cleaning up old mountid : done."
2024-06-27T16:05:08.9380267Z     daemon.go:317: [df78e0f4df204] failed to start daemon: error initializing graphdriver: driver not supported: overlay2
2024-06-27T16:05:08.9384340Z     prune_test.go:152: [df78e0f4df204] failed to start daemon with arguments [--data-root /go/src/github.com/docker/docker/bundles/test-integration/TestPruneDontDeleteUsedImage/single_tag/repo_and_digest,_no_tag/df78e0f4df204/root --exec-root /tmp/dxr/df78e0f4df204 --pidfile /go/src/github.com/docker/docker/bundles/test-integration/TestPruneDontDeleteUsedImage/single_tag/repo_and_digest,_no_tag/df78e0f4df204/docker.pid --userland-proxy=true --containerd-namespace df78e0f4df204 --containerd-plugins-namespace df78e0f4df204p --containerd /var/run/docker/containerd/containerd.sock --host unix:///tmp/docker-integration/df78e0f4df204.sock --debug --storage-driver overlay2] : [df78e0f4df204] daemon exited during startup: exit status 1
```

They work fine locally 🤔

laurazard commented:
I was testing this on graphdrivers vs containerd, and noticed on graphdrivers, we seem to always remove the last tag alphanumerically, but with the containerd implementation it seems a bit random.

I wrote a repro here:

```bash
#!/usr/bin/env bash
set -ex
while true; do
    docker image prune -af
    docker pull busybox:latest
    repoDigest=$(docker inspect busybox:latest | jq .[0].RepoDigests[0] | tr -d '"')
    docker tag busybox:latest busybox:a
    docker tag busybox:latest busybox:b
    docker tag busybox:latest busybox:c
    docker rmi busybox:latest
    ctrId=$(docker run -d $repoDigest sleep 100)
    yes | docker image prune -a
    tags=$(docker images --filter=reference='busy*' --format json | jq .Tag | tr -d '"')
    tagsLength=${#tags[@]}
    if [[ $tagsLength -ne 1 ]]; then
        echo "this shouldn't happen! tags length != 1"
        exit 1
    fi
    if [[ $tags != c ]]; then
        exit 1
    fi
    docker kill $ctrId
    docker rm $ctrId
done
```

laurazard commented:
I don't know whether **must** keep this behavior, but it sounds safer 😅.

vvoland committed:
c8d/prune: Keep deletion order stable

When untagging multiple images targetting the same digest, delete the
images in lexographic order to be consistent with graphdrivers.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented on integration/image/prune_test.go:
```diff
@@ -47,3 +52,172 @@ func TestPruneDontDeleteUsedDangling(t *testing.T) {
+
+	assert.Check(t, is.Len(pruned.ImagesDeleted, 1))
+	assert.Check(t, is.Equal(pruned.ImagesDeleted[0].Deleted, ""))
+	assert.Check(t, is.Equal(pruned.ImagesDeleted[0].Untagged, "busybox:a"))
```
Changed to 10 images tagged in non-lexographic order :)

laurazard reviewed (approved):
LGTM 🎉

thaJeztah commented on daemon/containerd/image_prune.go:
```diff
@@ -107,8 +108,16 @@ func (i *ImageService) pruneUnused(ctx context.Context, filterFunc imageFilterFu
 
+	// Sort images by name to make the behavior deterministic and consistent with graphdrivers.
+	sorted := make([]string, 0, len(imagesToPrune))
+	for name := range imagesToPrune {
+		sorted = append(sorted, name)
+	}
```
We can probably start looking at the `maps` package, and for these kind of things, use [`maps.Keys()`](https://pkg.go.dev/golang.org/x/exp/maps#Keys)

thaJeztah reviewed (approved):
LGTM, thanks!
